### PR TITLE
Handle PanicSentinel in Interpreter

### LINKED ISSUE #1412: Implement Dataflow Error Handling in the Execution Interpreter
### Summary

The Runtime has a special value (a Sentry) for the Dataflow Errors. We want to treat it in a special way in the interpreter, allowing us to continue the program execution in case of dataflow errors.

### Value

Better error reporting in the IDE.

### Specification

- [ ] Ping-pong the sentry in the `IdExecutionInstrument`
    - When the runtime encounters a sentry (value) it throws a SentryException
    - `IdExecutionInstrument#onException` catches the SentryException, 
      calls an appropriate error callback that notifies about the dataflow error
      (a SentryException should have an external id or a location of the failed expression),
      and tries to continue execution, by returning another sentry with an extended stack trace

### Acceptance Criteria & Test Cases

- In case of dataflow errors, the interpreter continues execution.

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #1412 

PR adds the ability to handle runtime exceptions in the interpreter and continue the execution passing the `PanicSentinel` value.

Changelog:
- update: `IdExecutionIstrument` logic to handle runtime panics
- refactor: API, remove poisoning message

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: distribution/std-lib/Base/src/Data/Number/Extensions.enso [hunk 1]
```diff
@@ -102,3 +103,16 @@ Decimal.parse : Text -> Decimal | Nothing
...
+Integer.parse text =
+    Panic.recover (Java_Integer.parseInt [text]) . catch (_ -> Nothing)
+
+## Parses a textual representation of a number into an integer or decimal number.
+   Returns `Nothing` if the text does not represent a valid number.
+Number.parse : Text -> Number | Nothing
+Number.parse text =
+    case Integer.parse text of
+        Nothing -> Decimal.parse text
+        value   -> value
```
Comment (iamrecursion) [inline distribution/std-lib/Base/src/Data/Number/Extensions.enso]: These, and `Decimal.parse` should use the dataflow errors mechanism, and return an `Unparseable` (or similar) error.

File: engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/ExpressionErrorsTest.scala [hunk 2]
```diff
@@ -141,16 +174,15 @@ class ExpressionErrorsTest
     val Some(Api.Response(_, Api.InitializedNotification())) = context.receive
   }
 
-  it should "return dataflow errors in method body" in {
+  it should "return panic sentinels in method body" in {
```
Comment (iamrecursion) [inline engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/ExpressionErrorsTest.scala:177]: You should have a test for dataflow errors as well.

File: test/Tests/src/Data/Numbers_Spec.enso [hunk 3]
```diff
@@ -190,4 +194,10 @@ spec =
             1.ceil . should_equal 1
             almost_max_long_times_three_decimal.ceil.to_decimal . should_equal almost_max_long_times_three_plus_1.to_decimal
             almost_max_long_times_three_plus_1.ceil . should_equal almost_max_long_times_three_plus_1
-
+        Test.specify "should parse text as decimals" <|
+           result = [-1.0, 0.0, 0.0, 1.0, 2.0, 3.4, Nothing]
+           text_literals.map Decimal.parse . should_equal result
+    Test.group "Numbers" <|
+        Test.specify "should parse text as numbers" <|
+           result = [-1, 0, 0, 1.0, 2.0, 3.4, Nothing]
```
Comment (iamrecursion) [inline test/Tests/src/Data/Numbers_Spec.enso]: This is very odd behaviour. If it is a valid integer it should be _parsed_ as a valid integer. The input list should result in `[-1, 0, 0, 1, 2, 3.4, Parse_Error]`. Basically, it should parse the numbers based on the natural representation for their value in our unified number system.

**Review by iamrecursion — CHANGES_REQUESTED** [review]
Good stuff, but a few things need changing.

File: engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/ExpressionErrorsTest.scala (hunk 2 above)
Comment (4e6) [reply engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/ExpressionErrorsTest.scala:177]: I added extra tests :+1:

File: distribution/std-lib/Base/src/Data/Number/Extensions.enso [hunk 4]
```diff
@@ -1,8 +1,12 @@
 from Base import all
 polyglot java import java.lang.Math
 polyglot java import java.lang.Double
+polyglot java import java.lang.Integer as Java_Integer
 polyglot java import java.lang.String
 
+## An error that occurs when the text does not represent a valid number.
+type Number_Parse_Error
```
Comment (iamrecursion) [inline distribution/std-lib/Base/src/Data/Number/Extensions.enso]: This should contain a description obtained from the Java parse error.

File: test/Tests/src/Data/Numbers_Spec.enso [hunk 5]
```diff
@@ -190,4 +198,20 @@ spec =
...
+            Decimal.parse "2.0" . to_text . should_equal "2.0"
+            Decimal.parse "3.4" . to_text . should_equal "3.4"
+            Test.expect_error_with (Decimal.parse "five") Number_Parse_Error
+    Test.group "Numbers" <|
+        Test.specify "should parse text as numbers" <|
+            Number.parse "-1" . to_text . should_equal "-1"
+            Number.parse "-0" . to_text . should_equal "0"
+            Number.parse "0" . to_text . should_equal "0"
+            Number.parse "1" . to_text . should_equal "1"
+            Number.parse "2.0" . to_text . should_equal "2.0"
```
Comment (iamrecursion) [inline test/Tests/src/Data/Numbers_Spec.enso]: Again, I don't think that this is implementing the correct behaviour. If a decimal is an exact integral value, then it should parse as an integer, not a decimal.

File: distribution/std-lib/Base/src/Data/Number/Extensions.enso [hunk 6]
```diff
@@ -98,7 +102,17 @@ Number.to_json = Json.Number this
...
+## Parses a textual representation of an integer into an integer number.
+   Returns `Nothing` if the text does not represent a valid integer.
+Integer.parse : Text -> Integer ! Number_Parse_Error
+Integer.parse text =
+    Panic.recover (Java_Integer.parseInt [text]) . catch (_ -> Error.throw Number_Parse_Error)
+
+## Parses a textual representation of a number into an integer or decimal number.
+   Returns `Nothing` if the text does not represent a valid number.
+Number.parse : Text -> Number ! Number_Parse_Error
+Number.parse text = Integer.parse text . catch (_ -> Decimal.parse text)
```
Comment (iamrecursion) [inline distribution/std-lib/Base/src/Data/Number/Extensions.enso]: Once you add descriptions to errors in this case it won't make sense here. You'll need to explicitly set the error message for this method failing.

**Review by iamrecursion — CHANGES_REQUESTED** [review]

**Review by iamrecursion — APPROVED** [review]

**Review by kustosz — APPROVED** [review]

**Pull request merged**